haddock-api/src/Haddock/GhcUtils.hs

```diff
@@ -499,3 +502,45 @@ tyCoFVsOfTypes' []       fv_cand in_scope acc = emptyFV fv_cand in_scope acc
 tyCoFVsBndr' :: TyVarBinder -> FV -> FV
 tyCoFVsBndr' (TvBndr tv _) fvs = FV.delFV tv fvs `unionFV` tyCoFVsOfType' (tyVarKind tv)
 
+
+-------------------------------------------------------------------------------
+-- * Defaulting RuntimeRep variables
+-------------------------------------------------------------------------------
+
+-- | Traverses the type, defaulting type variables of kind 'RuntimeRep' to
+-- 'LiftedType'. See 'defaultRuntimeRepVars' in IfaceType.hs the original such
+-- function working over `IfaceType`'s.
+defaultRuntimeRepVars :: Type -> Type
+defaultRuntimeRepVars = go emptyVarEnv
+  where
+    go :: TyVarEnv () -> Type -> Type
+    go subs (ForAllTy (TvBndr var flg) ty)
+      | isRuntimeRepVar var
+      , isInvisibleArgFlag flg
+      = let subs' = extendVarEnv subs var ()
+        in go subs' ty
+      | otherwise
+      = ForAllTy (TvBndr (updateTyVarKind (go subs) var) flg)
+                 (go subs ty)
+
+    go subs (TyVarTy tv)
+      | tv `elemVarEnv` subs
+      = TyConApp liftedRepDataConTyCon []
+      | otherwise
+      = TyVarTy (updateTyVarKind (go subs) tv)
```

Comment:
I think it's unnecessary to recurse into `tv`'s kind here. (At least, `IfaceType.defaultRuntimeRepVars` doesn't do this.)

Reply:
It is necessary, since we do recurse into type variable's kinds when computing the inferred order of type variables (in `tyCoFVsBndr'`). Perhaps we should not recurse into the type variables kind there?

Reply:
Upon further thought, maybe it's better if we do recurse here.

For pretty-printing purposes, it's probably overkill, since if you have a type like:

```
forall (r :: RuntimeRep) (a :: TYPE r). Int -> a
```

Then after defaulting, it'll still display as:

```
forall a. Int -> a
```

Regardless of whether you recurse into the kind of `a` (in `Int -> a`) or not. On the other hand, if you _don't_ recurse into the kind of `a`, then you have a strange situation where the kind of `a` in `forall a` is `Type`, but the kind of `a` in `Int -> a` is `TYPE r`. I suppose the only way that could ever really bite you is if you were to run Core Lint on these types (which I certainly don't imagine happens today), but for the sake of consistency, it might be best just to update the kind of `a` in `Int -> a` anyways.